statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,699 @@
+
+import statsmodels.tsa.base.tsa_model as tsbase
+
+def mat_sqrt(_2darray):
+    """Calculates the square root of a matrix.
+
+    Parameters
+    ----------
+    _2darray : ndarray
+        A 2-dimensional ndarray representing a square matrix.
+
+    Returns
+    -------
+    result : ndarray
+        Square root of the matrix given as function argument.
+    """
+    u_, s_, v_ = svd(_2darray, full_matrices=False)
+    s_ = np.sqrt(s_)
+    return chain_dot(u_, np.diag(s_), v_)
+
+def _endog_matrices(endog_tot, diff_lags, deterministic):
+    """Returns different matrices needed for parameter estimation. These consist
+    of elements of the data as well as elements representing deterministic
+    terms. A tuple of consisting of these matrices is returned.
+
+    Parameters
+    ----------
+    endog_tot : ndarray (neqs x total_nobs)
+        The whole sample including the presample.
+    diff_lags : int
+        Number of lags in the VEC representation.
+    deterministic, str {"", "co", "cc", "lt", "s"}
+            "" - no deterministic terms
+            "co" - constant outside the cointegration relation
+            "cc" - constant within the cointegration relation
+            "lt" - linear trend
+            "s" - seasonal terms
+            Combinations of these are possible (e.g. "cclt" or "colt" for
+            linear trend with intercept)
+
+    Returns
+    -------
+    y_1_T : ndarray (neqs x nobs)
+        The (transposed) data without the presample.
+        .. math:: (y_1, \ldots, y_T)
+    delta_y_1_T : ndarray (neqs x nobs)
+        .. math:: (y_1, \ldots, y_T) - (y_0, \ldots, y_{T-1})
+    y_min1 : ndarray (neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+        .. math:: (y_0, \ldots, y_{T-1}
+    delta_x : ndarray (diff_lags*neqs x nobs)
+        (dimensions assuming no deterministic terms are given)
+
+    References
+    ----------
+    [Lut2005]_, p. 286
+    """
+    # p. 286:
+    p = diff_lags+1
+    y = endog_tot
+    K = y.shape[0]
+    y_1_T = y[:, p:]
+    T = y_1_T.shape[1]
+    delta_y = np.diff(y)
+    delta_y_1_T = delta_y[:, p-1:]
+    y_min1 = y[:, p-1:-1]
+    # if "co" in deterministic and "cc" in deterministic:
+    #     raise ValueError("Both 'co' and 'cc' as deterministic terms given. " +
+    #                      "Please choose one of the two.")
+    # todo: optimize the following with np.row_stack()
+
+    if "cc" in deterministic:  # pp. 257, 299, 306, 307
+        # y_min1_mean = y_min1.mean(1)
+        y_min1 = vstack((y_min1,
+                         np.ones(T)))
+        # H = vstack((np.identity(K),
+        #             - y_min1_mean))
+        # y_min1 = H.T.dot(y_min1)
+
+    # the following gives the same result as the R-package tsDyn when the
+    # linear term is outside the cointegration relation:
+    # if "lt" in deterministic:  # p. 299
+    #     y_min1 = vstack((y_min1,
+    #                      np.arange(T)))
```

Comment:
we need this case also where linear trend is in the cointegrating equations.
AFAICS, you only have it in the diff equation